Add and test skip_times option for Predictor

bwspenc opened the pull request:
closes #6506

bwspenc committed:
Add option to skip specified times in predictors ref #6506

friedmud commented:
Since this is a new feature in the framework - we really should have a test in the framework for it.  Can you come up with a simple one?  Maybe have a jump in a BC on a transient diffusion calculation or something?

bwspenc commented:
@friedmud I'll do that.  I think the best test for this is to check what the residual is right after the predictor is applied (rather than counting the nonlinear iterations, which the current test does). I'll see if I can cook up a postprocessor that lives in the MOOSE tests that does that.

bwspenc commented:
I just pushed a new version of the commit to add a set of framework tests for the predictor.  To test this more reliably, I added options to the Residual postprocessor to enable it to output the initial residual (either before or after applying preset BCs). That code is all tested by the new framework tests that I added for the predictor.

friedmud commented on framework/src/predictors/AdamsPredictor.C:
```diff
@@ -73,6 +73,10 @@ AdamsPredictor::apply(NumericVector<Number> & sln)
   // That means that history control must go here.
   historyControl();
+
+  if (!shouldApply())
```
I would rather have this handled by the framework calling `shouldApply()` and then just _not_ calling `Predictor::apply()` if it's `false`...

bwspenc replied:
That ```historyControl()``` wouldn't happen if it were done that way.  Based on the comment above that call, it sounds like we'd need to call a method on the predictor at the beginning of the time step to do what you suggest (which probably isn't terribly difficult).  Is that what we should do?

friedmud replied:
Maybe split this up... give it an `initialize()`, `execute()` (or `apply()`) and `finalize()` just like other objects in MOOSE?

`initialize()` could be called all the time... then `shouldApply()` will be checked... the last two only get called if `shouldApply()` returns true.

Just a suggestion.  If no one else thinks it matters then it's probably fine how it is ;-)

bwspenc replied:
That looks really easy to do. We could add a ```timestepSetup()``` method to the predictor that gets called from ```NonlinearSystem::onTimestepBegin()```, and move everything in ```AdamsPredictor::historyControl()``` into that method.  Once we did that, we could have the framework call ```shouldApply()``` and only call ```apply()``` if that's true.  I'll go ahead and do that unless I hear otherwise.

bwspenc replied:
It looks like our wires got crossed.  So is the name ```initialize()``` preferred to ```timestepSetup()```? Would it get called at the beginning of the timestep or when the predictor is about to be applied?  Either way works.

friedmud replied:
I like `timestepSetup()` better for now... let's go that direction.

We can add an `initialize()` later if its needed.  `timestepSetup()` is even more standard across MOOSE.

bwspenc committed:
Add tests for skip_times option in Predictor closes #6506

Modify existing test in solid_mechanics and add version of the
test to skip the predictor on a step.  Add similar framework
tests.
Also add and test options to specify the residual type in the
Residual postprocessor to allow us to more reliably test the predictor.

bwspenc commented:
The version I just pushed has the API changes for Predictor that I discussed with @friedmud in this thread.

permcody commented:
This looks ok to me but it touches a few base systems in MOOSE. I wouldn't mind having @andrsd or @friedmud review this as well.

andrsd commented:
This is fine with me.
